_posts/2021-03-15-forecasting-time-series-with-torch_3/forecasting_3.Rmd

```diff
@@ -0,0 +1,445 @@
+    
+    # return last states for all layers
+    # per layer, a single tensor for GRU, a list of 2 tensors for LSTM
+    x <- x[[2]]
+    x
+    
+  }
+  
+)
+```
+
+### Decoder
+
+In the decoder, too, the main component is an RNN. In contrast to previously-shown architectures, though, it does not just return a prediction. It also reports back the RNN's final state.
+
+```{r}
+decoder_module <- nn_module(
+  
+  initialize = function(type, input_size, hidden_size, num_layers = 1) {
+    
+    self$type <- type
+    
+    self$rnn <- if (self$type == "gru") {
+      nn_gru(
+        input_size = input_size,
+        hidden_size = hidden_size,
+        num_layers = num_layers,
+        batch_first = TRUE
+      )
+    } else {
+      nn_lstm(
+        input_size = input_size,
+        hidden_size = hidden_size,
+        num_layers = num_layers,
+        batch_first = TRUE
+      )
+    }
+    
+    self$linear <- nn_linear(hidden_size, 1)
+    
+  },
+  
+  forward = function(x, state) {
+    
+    # input to forward:
+    # x is (batch_size, 1, 1)
+    # state is (1, batch_size, hidden_size)
+    x <- self$rnn(x, state)
+    
+    # break up RNN return values
+    # output is (batch_size, 1, hidden_size)
+    # next_hidden is
+    c(output, next_hidden) %<-% x
+    
+    output <- output$squeeze(2)
+    output <- self$linear(output)
+    
+    list(output, next_hidden)
+    
+  }
+  
+)
+```
+
+### `seq2seq` module
+
+In `seq2seq` is where the action happens. The plan is to encode once, then call the decoder in a loop.
+
+If you look back to decoder `forward()`, you see that it takes two arguments: `x` and `state`.
+
+Dependent on the context, `x` corresponds to one of three things.
+
+-   The very first time the decoder is called on an input sequence, `x` maps to the final input value. This is different from a task like machine translation, where you would pass in a start token. With time series, though, we'd like to continue where the actual measurements stop.
+
+-   In further calls, we want the decoder to continue from its most recent prediction. It is only logical, thus, to pass back the preceding forecast.
+
+-   That said, in NLP a technique called "teacher forcing" is commonly used to speed up training. With teacher forcing, instead of the forecast we pass the actual ground truth, the thing the decoder should have predicted. We do that only in a configurable fraction of cases, and -- naturally -- only while training. The rationale behind this technique is that without this form of re-calibration, consecutive prediction errors can quickly erase any remaining signal.
+
+`state`, too, is polyvalent. But here, there are just two possibilities.
+
+-   The first time the decoder is called, it is "seeded" with the final state from the encoder. Note how this is *the only time* we make use of the encoding.
+
+-   From then on, the decoder's own previous state will be passed. Remember how it returns two values, forecast and state?
```

Comment:
Again, you want what's called "list agreement" with your bullets. If you're giving us the two possibilities, give them to us first, and then expand on the idea.

Reply:
yup. Going for a compromise :-) (keeping some "deviance" ;-))

>  But here, there are just two possibilities: encoder state and decoder state.